### PR TITLE
State Machine Design #1

### PR DESCRIPTION
**Pull request by chatton**
### All Submissions:

This PR introduces the State Machine struct that will be used in a future PR to refactor our current reconciliation loop design.

Individual states are defined with a Reconcile function and they are then linked together using the `AddTransition` function. 

A boolean value is returned from these states to indicate if they have been completed, if not, they will be re-run in the next reconciliation. If they are compete, the next state will be executed in the next reconciliation. 

* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

**Comment by irajdeep** [comment]
Looks Nice, I had one high-level question (before digging into the code), is there any mechanism present to (check &)avoid a loop among the different states.

**Comment by chatton** [comment]
@irajdeep there nothing to prevent cycles within states (a cycle is used to handle scaling)

UpdateSts -> UpdateAutomationConfig -> UpdateStatus -> UpdateSts etc. as only a single member can be updated at a time.

However I think it makes sense to detect undesired cycles in some way and to essentially reset back to the first state if that is the case.

**Comment by irajdeep** [comment]
> However I think it makes sense to detect undesired cycles in some way and to essentially reset back to the first state if that is the case.

Exactly, I am a bit wary about ending in this situation unknowingly and perform an indefinite reconciliation.

**Comment by chatton** [comment]
@irajdeep maybe we can discuss this and create a follow up PR to address potential undesired cycles (we won't refactor any of the logic to use this new type until that is addressed)

**Comment by irajdeep** [comment]
> maybe we can discuss this and create a follow up PR to address potential undesired cycles (we won't refactor any of the logic to use this new type until that is addressed)

SGTM

**Review by irajdeep — COMMENTED** [review]
Overall it looks pretty good to me. Added some comments -- let me know your thoughts.

File: pkg/util/state/statemachine.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// State should provide a unique name, and a Reconcile function.
+// This function gets called by the Machine. The first two returned values
+// are returned to the caller, while the 3rd value is used to indicate if the
+// State completed successfully. A value of true will move onto the next State,
+// a value of false will repeat this State until true is returned.
+type State struct {
+	Name      string
```
[the remark this answers is not in the document]
Comment (chatton) [reply pkg/util/state/statemachine.go:16]: I chose to make it a string as this struct can potentially be used with enterprise as well.

File: pkg/util/state/statemachine.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	Reconcile func() (reconcile.Result, error, bool)
+	OnEnter   func() error
```
[the remark this answers is not in the document]
Comment (chatton) [reply pkg/util/state/statemachine.go:18]: I'll add some additional docstrings!

File: pkg/util/state/statemachine.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+// transition represents a transition between two states.
+type transition struct {
+	from, to  State
+	predicate TransitionPredicate
```
[the remark this answers is not in the document]
Comment (chatton) [reply pkg/util/state/statemachine.go:24]: The predicate allows us to define what a valid transition is, for example, we want to transition from "CreateServce" -> "EnsureTLSResources" only when tls is enabled on the resource. We can do this with something like

```golang
sm.AddTransition(createServce, ensureTlsResources, func() bool {
 return mdb.Spec.Security.Tls.Enabled
})
```
We build up the StateMachine with each reconciliation, but the transitions that are valid are dependant on the current state of the resource.

[the remark this answers is not in the document]
Comment (irajdeep) [reply pkg/util/state/statemachine.go:24]: ack

**Review by irajdeep — APPROVED** [review]
Looks Great 👍

**Pull request merged**